trieste/utils/mo_objectives.py

```diff
@@ -0,0 +1,161 @@
+# Copyright 2020 The Trieste Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains toy multi-objective functions, useful for experimentation.
+"""
+from __future__ import annotations
+
+import math
+from abc import ABC, abstractmethod
+from functools import partial
+
+import tensorflow as tf
+
+from ..type import TensorType
+
+
+class MultiObjectiveTestProblem(ABC):
+    """base class for synthetic multi-objective test functions"""
+
+    @abstractmethod
+    def prepare_benchmark(self):
+        """
+        Evaluate original problem.
+        """
+
+    @abstractmethod
+    def gen_pareto_optimal_points(self, n: int):
+        """
+        Generate `n` pareto optimal points.
+        """
+
+
+class VLMOP2(MultiObjectiveTestProblem):
+    """
+    The VLMOP2 function, typically evaluated over :math:`[-2, 2]^2`. See
+    See :cite:`van1999multiobjective` for details.
+    """
+
+    bounds = [[-2.0] * 2, [2.0] * 2]
```

Comment:
Are bounds and dim supposed to be a required part of every MultiObjectiveTestProblem? If so they should be specified and documented in the ABC to ensure that everything implements them (you may need to mark them as @abstractmethod @property to do this).

Reply:
Is it possible to maybe elaborate more on this point?
The things are for some synthetic function the dim and bounds are used as a class variable as one could specify it in advance, while for some it needs to be specified during instantiation as input arg. 
Is that mean we would like them to be as property with some `getter` and `setter`?

Reply:
I meant that if all classes have a dim and bounds then we should specify and document this in the MultiObjectiveTestProblem ABC. But Python doesn't let you add abstract attributes, only abstract methods or properties. So you could have:

````python
class MultiObjectiveTestProblem(ABC):

    @property
    @abstractmethod
    def dim(self) -> int:
        """ The dimension """
````

And then the classes could implement it, either statically:

```python
class VLMOP2(MultiObjectiveTestProblem):
    dim = 2
````

Or dynamically:

````python
class DTLZ(MultiObjectiveTestProblem):
    def __init__(self, input_dim: int, ...):
        ....
        self._dim = input_dim

    @property
    def val(self) -> int:
        return self._dim
````

Reply:
Many thanks! I am not totally sure if the `bounds` is a must need as well as `dim`, but for now think it is good to add as well.